tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,272 @@
+
+Thus for the cell at $i^{th}$ index in the topmost row, it's value is multiplied by $n - 1 \choose i$ and added to the final sum $modulo\,10$.
+Time Complexity of the program is $O(n^2)$ for computing the binomial coefficient and $O(n)$ Space complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int triangularSum(vector<int>& nums) {
+        int n = nums.size();
+        vector<int> pascalTriangleRow = {1};
+        // Calculate the ith row using (i - 1)th row
+        for(int i = 0; i < n; i++){
+            vector<int> nextRow = {1};
+            for(int j = 1; j < i; j++){
+                nextRow.push_back((pascalTriangleRow[j] + pascalTriangleRow[j - 1])%10);
+            }
+            nextRow.push_back(1);
+            pascalTriangleRow = nextRow;
+        }
+        // Calculate the final answer as discussed above
+        int ans = 0;
+        for(int i = 0; i < n; i++){
+            ans += (nums[i]*pascalTriangleRow[i])%10;
+        }
+        return ans%10;
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+Sometimes it is not possible to calculate the entirety of Pascal's Triangle due to larger values of $n$. In this case, we begin by precomputing $x!$   $\forall$ $x \in [{0, n}]$. Similarly, we will also [precompute](https://cp-algorithms.com/algebra/module-inverse.html#mod-inv-all-num) the modular inverses. This can be achieved in $O(n)$ time. Thus we can now compute $n \choose r$ using the analytical equation presented earlier. You can read about modular inverses [here](../../tutorials/basic-topics/mod.md)
+
+The implementation of above can be as follows:-
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// Set the maximum value of n
+int N = 1e5;
+int mod = 1e9 + 7;
+
+// Making arrays to store the factorial and inverse factorial modulo m
+vector<long long int> factorial;
+vector<long long int> inverse_factorial;
+
+// Function to calculate nCr(n, r)
+long long int nCr(int n, int r){
+    if((r < 0) || (r > n)) return 0;
+    return ((factorial[n] * inverse_factorial[r])%mod * inverse_factorial[n - r])%mod;
+}
+
+int main(){
+    // Calculate values for factorial
+    factorial.push_back(1);
+    for(int i = 1; i <= N; i++) factorial.push_back((factorial.back()*i)%mod);
+
+    // Calculate values for inverse factorial
+    // For explanation and proof of this method, visit https://cp-algorithms.com/algebra/module-inverse.html#mod-inv-all-num
+    vector<long long int> inverse;
+    inverse.push_back(1);
+    inverse.push_back(1);
+    inverse_factorial.push_back(1);
+    inverse_factorial.push_back(1);
+    for(int i = 2; i <= N; i++){
+        inverse.push_back((mod - ((mod/i) * inverse[mod%i])%mod)%mod);
+        inverse_factorial.push_back((inverse_factorial[i - 1] * inverse[i])%mod);
+    }
+
+    cout << nCr(5, 0) << "\n";
+
+    return 0;
```

Comment:
the implementation is fine but it is better to make it as a template, so that people can just use it directly.

e.g.

```cpp
struct comb {
 public:
  vector<long long> fac, finv, inv;
  comb(int mxN) {
    fac.resize(mxN);
    finv.resize(mxN);
    inv.resize(mxN);
    fac[0] = fac[1] = 1;
    finv[0] = finv[1] = 1;
    inv[1] = 1;
    for (int i = 1; i < mxN - 1; i++) {
      fac[i + 1] = fac[i] * (i + 1) % MOD;
      inv[i + 1] = MOD - inv[MOD % (i + 1)] * (MOD / (i + 1)) % MOD;
      finv[i + 1] = finv[i] * inv[i + 1] % MOD;
    }
  }

  long long ncr(long long n, long long k) {
    if (k < 0 || n < k) return 0;
    long long res = 1;
    res *= fac[n] * finv[k] % MOD * finv[n - k] % MOD;
    return res;
  }
};
```

```cpp
comb c = comb(n);
cout << c.ncr(n, r) << endl;
```